WIP: First cut on clipped factorization analysis

timodonnell opened the pull request:
Another WIP on a new analysis method. Not ready for merging but feedback welcome.

laserson commented on phip/clipped_factorization_analysis.py:
```diff
@@ -0,0 +1,144 @@
+def do_clipped_factorization_analysis(
+        counts_df,
+        rank=3,
+        truncate_percentile=99.9,
```
How can we determine what the best value for this is?

timodonnell replied:
We should be pretty robust to this choice (e.g. 99 vs. 99.9 vs. 99.99). It just needs to be a lower-bound on fraction of non-hits. May be good to try different values though as we analyze the next few data sets

laserson commented on phip/clipped_factorization_analysis.py:
```diff
@@ -0,0 +1,144 @@
+
+    # Derived quantities
+    reconstruction = tf.matmul(
+        tf.gather(a, minibatch_indices), b)
```
Is there an implication for sampling rows but not cols? Or is this basically just a pragmatic issue?

timodonnell replied:
Yeah taking the minibatches row-wise is really just a pragmatic thing, since that's the long dimension in our data

laserson commented on phip/clipped_factorization_analysis.py:
```diff
@@ -0,0 +1,144 @@
+    last_log_at = 0
+    with tf.Session() as session:
+        session.run(init)
+        all_indices = numpy.array(list(range(reads_per_million.shape[0])))
```
`all_indices = numpy.arange(reads_per_million.shape[0])`?

timodonnell replied:
👍

laserson commented on phip/clipped_factorization_analysis.py:
```diff
@@ -0,0 +1,144 @@
+import time
+
+import pandas
```
I'd go with either

```
import pandas as pd
```

or

```
from pandas import DataFrame
```

timodonnell replied:
👍

laserson commented on phip/clipped_factorization_analysis.py:
```diff
@@ -0,0 +1,144 @@
+import time
+
+import pandas
+import numpy
```
same...I guess `import numpy as np`

timodonnell replied:
👍

laserson commented on phip/clipped_factorization_analysis.py:
```diff
@@ -0,0 +1,144 @@
+        tf.pow(differences, 2), axis=1)
+    loss = (
+        tf.reduce_mean(
+            tf.pow(tf.minimum(differences, truncate_threshold), 2))
```
Strictly speaking, is this the Frobenius norm? Don't you need to sum the matrix elts and take the square root? This would change the values of the gradients, no?

timodonnell replied:
You're correct, but while it will change the value of the gradients it will not change the coordinates of the minimum point, so we are still optimizing the advertised function. Taking the mean instead of sum is mean to keep the loss roughly on a similar scale to the data entries (so e.g. learning rate isn't tightly dependent on matrix size) and skipping the square root is for speed.

timodonnell commented:
Thanks for the comments @laserson ! Updated to address comments and rebase onto current master. After playing with this for a while I think it's reasonable to merge after we address any remaining things you notice. Additional changes I'm making now:

* Added unit test
* Rename all occurrences of "truncated factorization" to "clipped factorization"
* Better docstring
* Small tweaks to string formatting so as not to require python 3.6

laserson commented:
lgtm! feel free to merge, but I'd do a rebase onto master beforehand just in case.
